slp/modules/rnn.py

```diff
@@ -9,29 +9,33 @@
 class RNN(nn.Module):
     def __init__(self, input_size, hidden_size, batch_first=True,
                  layers=1, bidirectional=False, dropout=0,
-                 rnn_type='lstm', packed_sequence=True):
+                 rnn_type='lstm', packed_sequence=True, device='cuda'):
         super(RNN, self).__init__()
         rnn_cls = nn.LSTM if rnn_type == 'lstm' else nn.GRU
         self.rnn = rnn_cls(input_size,
                            hidden_size,
-                           batch_first=False,
+                           batch_first=batch_first,
                            num_layers=layers,
                            bidirectional=bidirectional)
         self.drop = nn.Dropout(dropout)
-        self.packed_sequence = None
+        self.packed_sequence = packed_sequence
         if packed_sequence:
             self.pack = PackSequence(batch_first=batch_first)
             self.unpack = PadPackedSequence(batch_first=batch_first)
+        self.device = device
+        self.bidirectional = bidirectional
 
     def forward(self, x, lengths):
         self.rnn.flatten_parameters()
-        x, lengths = self.pack(x, lengths)
-        out, _ = self.rnn(x)
-        if self.packed_sequence is not None:
+        if self.packed_sequence:
+            x, lengths = self.pack(x, lengths)
+        out, (hidden, cell) = self.rnn(x)
+        if self.packed_sequence:
             out = self.unpack(out, lengths)
         out = self.drop(out)
-        last_hidden = out[:, -1, :]
-        return out, last_hidden
+        # ->>> last_hidden = out[:, -1, :]
+        last_hidden = torch.cat((hidden[-2,:,:], hidden[-1,:,:]), dim=1) if self.bidirectional else hidden[-1,:,:]
```

Comment:
@ckarouzos Does this work for batch_first=True?

Reply:
That is the case by default. 
However it's not tested for batch_first=False.